diff --git a/src/itemset.c b/src/itemset.c
index d7ec189..b38ba7b 100644
--- a/src/itemset.c
+++ b/src/itemset.c
@@ -117,8 +117,6 @@ itemset_generic_merge_check (GList *items, itemPtr newItem, gint maxChecks, gboo
 				found = TRUE;
 
 				/* found corresponding item, check if they are REALLY equal (eg, read status may have changed) */
-				if(oldItem->readStatus != newItem->readStatus) 
-					equal = FALSE;
 				if(oldItem->flagStatus != newItem->flagStatus)
 					equal = FALSE;
 				break;
